File: Difficulty: Medium/Next element with greater frequency/next-element-with-greater-frequency.cpp
class Solution {
  public:
    vector<int> findGreater(vector<int>& arr) {
        int n = arr.size();
        vector<int> ans(n, -1);
        
        unordered_map<int, int> freq;
        for(int num : arr) freq[num]++;
        
        stack<pair<int, int>> st; // <num, freq[num]>
        
        for(int i=n-1; i>=0; i--) {
            while(!st.empty() && st.top().second <= freq[arr[i]]) {
                st.pop();
            }
            
            if(!st.empty()) {
                ans[i] = st.top().first;
            }
            
            st.push({arr[i], freq[arr[i]]});
        }
        
        return ans;
    }
};

File: Difficulty: Medium/Next element with greater frequency/next-element-with-greater-frequency.java
class Solution {
    public ArrayList<Integer> nextFreqGreater(int[] arr) {
        int max = Integer.MIN_VALUE;
        for(int i=0; i<arr.length; i++)
            if(max < arr[i])
                max = arr[i];
        
        int[] num = new int[max+1];
        for(int i=0; i<arr.length; i++)
            num[arr[i]]++;
        
        ArrayList<Integer> list = new ArrayList<>();
        for(int i=0; i<arr.length-1; i++) {
            boolean flag = true;
            for(int j=i+1; j<arr.length; j++) {
                if(num[arr[i]] < num[arr[j]]) {
                    list.add(arr[j]);
                    flag = false;
                    break;
                }
            }
            if(flag) list.add(-1);
        }
        
        list.add(-1);
        return list;
    }
}